### DISCUSSION TITLE
GetTopLevel() from ViewModel

### DISCUSSION THREAD

**Discussion by katterfelto** [AvaloniaUI/Avalonia#12475] [Q&A] [answered]
Hi 

I am using CommunityToolkit.MVVM with Avalonia v11. I am trying open an OpenFileDialog using StorageProvider from a RelayCommand function in my ViewModel class. 

What is the best way to get a reference to a control into the class/function so I can pass it into the GetTopLevel() function?

I am an experienced winforms developer but new to WPF/MVVM.

TIA
Dave

**Comment by thevortexcloud** [comment]
If you are interested in doing things properly, then you don't. The View Model should to the best of its ability have no direct connection to any kind of presentation logic (this makes switching to a different UI framework/writing unit tests significantly easier). If you were using ReactiveUI, you would use ReactiveUI's interactions (it's sort of like a blocking event that allows for data to go two ways) to run code inside the view class which can then open the dialogue for you and then pass a result back to the view model. But I am not familiar with the CommunityMVVM way of doing it. There is examples for doing the ReactiveUI way on Avalonia's doc site.

If you don't care about doing it properly/want a shortcut you can technically have the view call the viewmodel via the view's data context property which should be set to an instance of the view model. But again, you probably should not do that.

**Comment by timunie** [comment, marked as answer]
@katterfelto I've a sample in preview here: https://github.com/timunie/Avalonia.Samples/tree/feature/MVVM_Dialogs/src/Avalonia.Samples/MVVM/AdvancedMvvmDialogSample 

Note: it's not yet finished and not merged yet, but in general it should do what you need. Hope it helps you.

**Reply by katterfelto** [reply]
@timunie thanks for the pointer, I have figured out what it is doing and it is so close to being perfect for me. 

I integrated the Services folder from that sample into my app and everything seems good except for the `DialogService` constructor. I get a compilation error on `RegisterProperty.Changed.Subscribe(RegisterChanged);`. If I add back in the `Avalonia.ReactiveUI` NuGet package to my project this error goes away. it must have a extension method to cope with this parameter type.

**Reply by timunie** [reply]
I thought System.Readtive should be enough for this to work. Will investigate it in the sample. Thanks for the feedback 👍

**Reply by katterfelto** [reply]
You are correct, once I deciphered the auto-correct 😊

**Reply by Koushakur** [reply]
This sounds like what I'm looking for but the sample does no longer exist, just leads to a 404 page

**Reply by thevortexcloud** [reply]
That's because it's here now: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/ViewInteraction

**Comment by workgroupengineering** [comment]
I am using the following approach:

```csharp
    public interface IViewToViewModelBridge
    {
        Visual? View { get; set; }
    }

    public class ViewModelBase : INotifyPropertyChanged, IViewToViewModelBridge
    {
        private readonly WeakReference<Visual?> _visualRef = new WeakReference<Visual?>(default);
        Visual? IViewToViewModelBridge.View 
        { 
            get => _visualRef.TryGetTarget( out var target) ? target : default;
            set => _visualRef.SetTarget(value);
        }

        public Avalonia.Controls.TopLevel? TopLevel
        {
            get
            {
                if (_visualRef.TryGetTarget(out var view) && view is not null)
                {
                    return Avalonia.Controls.TopLevel.GetTopLevel(view);
                }
                return default;
            }
        }

        public event PropertyChangedEventHandler? PropertyChanged;
        protected bool RaiseAndSetIfChanged<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
        {
            if (!EqualityComparer<T>.Default.Equals(field, value))
            {
                field = value;
                RaisePropertyChanged(propertyName);
                return true;
            }
            return false;
        }


        protected void RaisePropertyChanged([CallerMemberName] string? propertyName = null)
            => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    public class ViewLocator : Avalonia.Controls.Templates.IDataTemplate
    {
        public Avalonia.Controls.Control? Build(object? data)
        {
            if (data is null)
                return null;

            var name = data.GetType().FullName!.Replace("ViewModel", "View");
            var type = Type.GetType(name);

            Control? view = default;

            if (type != null)
            {
                view = (Avalonia.Controls.Control)Activator.CreateInstance(type)!;
            }
            else
            {
                view = new Avalonia.Controls.TextBlock { Text = name };
            }

            if (data is IViewToViewModelBridge bridge)
            {
                bridge.View = view;
            }

            return view;
        }

        public bool Match(object? data)
        {
            return data is ViewModelBase;
        }
    }
```

and using:

```c#
    public class MyViewModel : ViewModelBase
    {

        public bool CanOpenFile(object? parameter) =>
            TopLevel?.StorageProvider is Avalonia.Platform.Storage.IStorageProvider { CanOpen: true };

        public async void OpenFile(object? parameter)
        {
            if (TopLevel?.StorageProvider is Avalonia.Platform.Storage.IStorageProvider { CanOpen: true } provider)
            {
                await provider.OpenFilePickerAsync(new());
            }
            else
                throw new NotSupportedException();
        }
    }
```

**Comment by Coloryr** [comment]
I'm use this
```
public partial class BaseModel : ObservableObject
{
    protected readonly IUserControl Control;

    public IBaseWindow Window => Control.Window;

    public BaseModel(IUserControl con)
    {
        Control = con;
    }
}
```
```
public partial class TestControl : UserControl, IUserControl
{
    public Window Window => App.FindRoot(VisualRoot);
    public UserControl Con => this;
    private readonly BaseModel _model;

    public TestControl()
    {
        InitializeComponent();

        _model = new(this);
        DataContext = _model;
    }
}
```

```
public interface IUserControl
{
    public UserControl Con { get; }
    public Window Window { get; }
}
```

**Comment by cirrusone** [comment]
There are some excellent examples here. 

If you prefer the `MVVM Community Toolkit` flavour then there are also some dialog examples here https://github.com/CommunityToolkit/MVVM-Samples

**Comment by katterfelto** [comment]
Thanks everyone, sorry for the slow response.

This got me staggering forward again and lost of new examples to study.
